File: Pa1_Files/List.java
/**
 * Name: Garland Simpson
 * Cruzid:
 * Assignment: pa1
 */

public class List
{
    private int n;
    private int index;
    private Node cursor;
    private Node front;
    private Node back;

    // Constructor: Creates a new empty list.
    List()
    {
        n = 0;
        index = -1;
        cursor = null;
        front = null;
        back = null;
    }


    private class Node
    {
        private int element;
        private Node next;
        private Node previous;

        Node(int theElement)
        {
            element = theElement;
            next = previous = null;
        }
    }

    // Access functions
    int length()//Returns the number of elements in this List.
    {
        return n;
    }

    int index() //If cursor is defined, returns the index of the cursor element, otherwise returns -1.
    {
        return index;
    }

    int front() // Returns front element. Pre: length()>0
    {
        if (n <= 0)
            throw new RuntimeException("Error in front(): length <= 0");
        else
            return front.element;
    }

    int back()  // Returns back element. Pre: length()>0
    {
        if (n <= 0)
            throw new RuntimeException("Error in back(): length <= 0");
        else
            return back.element;
    }

    int get()  // Returns cursor element. Pre: length()>0, index()>=0
    {
        if (n <= 0)
            throw new RuntimeException("Error in get(): length <= 0");
        else if (index < 0)
            throw new RuntimeException("Error in get(): index < 0");
        else
            return cursor.element;
    }

    public boolean equals(List L) // Returns true if this List and L are the same integer
    // sequence. The cursor is ignored in both lists.
    {
        if (n != L.n) return false; // CASE: RHS Length != LHS Length.

        Node current = front;
        Node lCurrent = L.front;

        while(current != null)
        {
            if (current.element != lCurrent.element) // CASE: the elements are NOT Equal.
                return false;

            current = current.next;
            lCurrent = lCurrent.next;
        }

        return true; // CASE: The Lists are the same.
    }


    // Manipulation procedures
    void clear() // Resets this List to its original empty state.
    {
        n = 0;
        index = -1;
        cursor = null;
        front = null;
        back = null;
    }

    void moveFront() // If List is non-empty, places the cursor under the front element, otherwise does nothing.
    {
        if (n != 0)
        {
            index = n - 1;
            cursor = front;
        }
    }

    void moveBack() // If List is non-empty, places the cursor under the back element, otherwise does nothing.
    {
        if (n != 0)
        {
            index = n - 1;
            cursor = back;
        }
    }

    void movePrev() // If cursor is defined and not at front, moves cursor one step toward
    // front of this List, if cursor is defined and at front, cursor becomes
    // undefined, if cursor is undefined does nothing.
    {
        if (cursor != null)
        {
            cursor = cursor.previous;
            index--;
        }
    }

    void moveNext() // If cursor is defined and not at back, moves cursor one step toward
    // back of this List, if cursor is defined and at back, cursor becomes
    // undefined, if cursor is undefined does nothing.
    {
        if (cursor != null)
        {
            cursor = cursor.next;
            index--;
        }
    }

    void prepend(int data) // Insert new element into this List.
    {
        Node newNode = new Node(data);

        if (n == 0) // CASE: list is empty.
        {
            newNode.next = null;
            newNode.previous = null;
            front = newNode;
            back = newNode;
            cursor = newNode;
            index = 0;
        }
        else // CASE: List is NOT empty, insert before front;
        {
            newNode.previous = null;
            newNode.next = front;
            front.previous = newNode;
            front = newNode;
        }
        n++;
    }

    void append(int data) // Insert new element into this List.
    {
        Node newNode = new Node(data);

        if (n == 0) // CASE: List is empty.
        {
            newNode.previous = null;
            newNode.next = null;
            front = newNode;
            back = newNode;
            cursor = newNode;
            index = 0;
        }
        else // CASE: List is NOT empty, insert after back.
        {
            newNode.next = null;
            newNode.previous = back;
            back.next = newNode;
            back = newNode;
        }
        n++;
    }

    void insertBefore(int data) // Insert new element before cursor. Pre: length()>0, index()>=0
    {
        if (n <= 0)
            throw new RuntimeException("Error in insertBefore(int): length <= 0");
        else if (index < 0)
            throw new RuntimeException("Error in insertBefore(int): index < 0");
        else
        {
            Node newNode = new Node(data);

            if (cursor.previous == null)
            {
                front = newNode;
                newNode.previous = null;
                newNode.next = cursor;
                cursor.previous = newNode;
            }
            else
            {
                newNode.previous = cursor.previous;
                newNode.next = cursor;
                cursor.previous = newNode;
                cursor = newNode.previous;
                cursor.next = newNode;
                cursor = newNode.next;
            }
            n++;
        }
    }

    void insertAfter(int data) // Inserts new element after cursor. Pre: length()>0, index()>=0
    {
        if (n <= 0)
            throw new RuntimeException("Error in insertAfter(int): length <= 0");
        else if (index < 0)
            throw new RuntimeException("Error in insertAfter(int): index < 0");
        else
        {
            Node newNode = new Node(data);

            if (cursor.next == null)
            {
                cursor.next = newNode;
                newNode.next = null;
                newNode.previous = cursor;
                back = newNode;
            }
            else
            {
                newNode.next = cursor.next;
                newNode.previous = cursor;
                cursor.next = newNode;
                cursor = newNode.next;
                cursor.previous = newNode;
                cursor = newNode.previous;
            }
            n++;
        }
    }

    void deleteFront() // Deletes the front element. Pre: length()>0
    {
        if (n <= 0)
            throw new RuntimeException("Error in deleteFront(): length <= 0");
        else
        {
            cursor = front;
            front = front.next;
            front.previous = null;
            cursor.next = null;
            cursor = front;
            n--;
        }
    }

    void deleteBack() // Deletes the back element. Pre: length()>0
    {
        if (n <= 0)
            throw new RuntimeException("Error in deleteBack(): length <= 0");
        else
        {
            cursor = back;
            back = back.previous;
            back.next = null;
            cursor.previous = null;
            cursor = back;
            n--;
        }
    }

    void delete() // Deletes cursor element, making cursor undefined. Pre: length()>0, index()>=0
    {
        if (n <= 0)
            throw new RuntimeException("Error in delete(): length <= 0");
        else if (index < 0)
            throw new RuntimeException("Error in delete(): index < 0");
        else
        {
            Node temp = cursor;
            cursor = temp.previous;
            cursor.next = temp.next;
            cursor = temp.next;
            cursor.previous = temp.previous;
            cursor = null;
            index = -1;
            n--;
        }
    }

    // Other methods
    @Override
    public String toString() // Overrides Object's toString method. Returns a String
    // representation of this List consisting of a space
    // separated sequence of integers, with front on left.
    {
        StringBuffer sb = new StringBuffer();
        if (n > 0)
        {
            Node current = front;

            while(current != null)
            {
                sb.append(current.element);

                if(current != back)
                    sb.append(" ");

                current = current.next;
            }
        }

        return new String(sb);
    }

    List copy() // Returns a new List representing the same integer sequence as this
    // List.  The cursor in the new list is undefined, regardless of the
    // state of the cursor in this List. This List is unchanged.
    {
        List newList = new List();
        Node current = front;

        while(current != null)
        {
            newList.append(current.element);
            current = current.next;
        }

        return newList;
    }
}

File: Pa4_files/List.c
// Name: Garland Simpson
// cruz-id: gfsimpso
// Assignment: pa4


#include<stdlib.h>
#include "List.h"

// structs --------------------------------------------------------------------

// private NodeObj type
typedef struct NodeObj {
	int element;
	struct NodeObj* next;
	struct NodeObj* prev;
} NodeObj;

// private Node type
typedef NodeObj* Node;

// private QueueObj type
typedef struct ListObj {
	Node front;
	Node back;
	Node cursor;
	int index;
	int n;
} ListObj;


// Constructors-Destructors ---------------------------------------------------

// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data) {
	Node N = malloc(sizeof(NodeObj));
	N->element = data;
	N->next = N->prev = NULL;
	return(N);
}

// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN) {
	if (pN != NULL && *pN != NULL) {
		free(*pN);
		*pN = NULL;
	}
}

// newQueue()
// Returns reference to new empty Queue object.
List newList(void) {
	List L;
	L = malloc(sizeof(ListObj));
	L->front = L->back = NULL;
	L->cursor = NULL;
	L->n = 0;
	L->index = -1;
	return(L);
}


void freeList(List* pL) {

	if (pL != NULL && *pL != NULL) {
		clear(*pL);
		free(*pL);
		pL = NULL;
	}
}

// Access functions -----------------------------------------------------------
int length(List L) {

	if (L == NULL) {

		printf("Queue Error: calling getLength() on NULL Queue reference\n");
		exit(1);
	}
	return(L->n);
}

int index(List L) {
	if (L->cursor == L->front && L->index != -1)
		return 0;

	return(L->index);
}

int front(List L) {

	if (L == NULL) {

		printf("Queue Error: calling front() on NULL List reference\n");
		exit(1);
	}
	return(L->front->element);
}

int back(List L) {

	if (L == NULL) {

		printf("Queue Error: calling back() on NULL List reference\n");
		exit(1);
	}
	return(L->back->element);
}

int get(List L) {

	if (L == NULL) {

		printf("Queue Error: calling back() on NULL List reference\n");
		exit(1);
	}
	return(L->cursor->element);
}

int equals(List A, List B) {

	if (A == NULL || B == NULL) {
		printf("List Error: calling equals() on NULL List reference\n");
		exit(1);
	}

	//CASE: List lengths not equal
	if (A->n != B->n) return 0;

	Node aTemp = A->front;
	Node bTemp = B->front;

	while (aTemp != NULL) {

		//CASE: the elements are not the same
		if (aTemp->element != bTemp->element) {

			aTemp = bTemp = NULL;
			return 0;
		}

		aTemp = aTemp->next;
		bTemp = bTemp->next;
	}

	//CASE: The lists are equal
	return 1;
}

// Manipulation procedures ----------------------------------------------------
void clear(List L) {

	while (length(L) > 0)
		deleteBack(L);
}

void moveFront(List L) {

	if (L->n != 0) {
		L->index = 0; // L->index = L->n - 1;
		L->cursor = L->front;
	}
}

void moveBack(List L) {

	if (L->n != 0) {
		L->index = L->n - 1;
		L->cursor = L->back;
	}
}

void movePrev(List L) {

	if (L->cursor != NULL) {
		L->index--;
		L->cursor = L->cursor->prev;
	}
	// If cursor is defined and not at front, moves cursor one step toward
	// front of this List, if cursor is defined and at front, cursor becomes
	// undefined, if cursor is undefined does nothing. 
}

void moveNext(List L) {

	if (L->cursor != NULL) {
		if (L->cursor == L->back) {
			L->index = -1;
			L->cursor = NULL;
		}
		else {
			L->index++;
			L->cursor = L->cursor->next;
		}

	}
}

void prepend(List L, int data) {

	Node N = malloc(sizeof(NodeObj));
	N->next = N->prev = NULL;
	N->element = data;

	//CASE: List is empty.
	if (L->front == NULL) {

		L->front = N;
		L->back = N;
		L->cursor = N;
		L->index = 0;
	}

	//CASE: List is NOT empty.
	else {
		L->front->prev = N;
		N->next = L->front;
		L->front = N;
		L->index++; // this is new

	}
	L->n++;
}

void append(List L, int data) {

	Node N;
	N = malloc(sizeof(NodeObj));
	N->next = NULL;
	N->prev = NULL;
	N->element = data;

	//CASE: List is empty.
	if (L->front == NULL) {

		L->front = N;
		L->back = N;
		L->cursor = N;
		L->index = 0;
	}

	//CASE: List is NOT empty.
	else {

		N->prev = L->back;
		L->back->next = N;
		L->back = N;
	}

	L->n++;
}

void insertBefore(List L, int data) {

	//CASE: List is NULL
	if (L == NULL) {
		printf("List Error: calling insertBefore() on NULL List reference\n");
		exit(1);
	}
	//CASE: List is empty
	if (L->n == 0) {
		printf("List Error: calling insertBefore() on an empty List\n");
		exit(1);
	}

	Node N = malloc(sizeof(NodeObj));

	N->next = NULL;
	N->prev = NULL;
	N->element = data;

	if (L->cursor->prev == NULL) {

		L->front = N;
		N->next = L->cursor;
		L->cursor->prev = N;
		L->index++;  // was L->index = 0;
	}
	else {
		N->prev = L->cursor->prev;
		N->next = L->cursor;
		L->cursor = N->prev;
		L->cursor->next = N;
		L->cursor = N->next;
		L->cursor->prev = N;
		L->index++;
	}
	L->n++;
}

void insertAfter(List L, int data) {

	//CASE: List is NULL
	if (L == NULL) {
		printf("List Error: calling insertAfter() on NULL List reference\n");
		exit(1);
	}
	//CASE: List is empty
	if (L->n == 0) {
		printf("List Error: calling insertAfter() on an empty List\n");
		exit(1);
	}

	Node N = malloc(sizeof(NodeObj));

	N->next = NULL;
	N->prev = NULL;
	N->element = data;

	if (L->cursor->next == NULL) {

		N->prev = L->cursor;
		L->cursor->next = N;
		L->back = N;
		L->index = 0;
	}
	else {
		N->next = L->cursor->next;
		L->cursor->next = N;
		N->prev = L->cursor;
		L->cursor = N->next;
		L->cursor->prev = N;
		L->cursor = N->prev;
	}
	L->n++;
}

void deleteFront(List L) {

	//CASE: List is NULL
	if (L == NULL) {
		printf("List Error: calling deleteFront() on NULL List reference\n");
		exit(1);
	}
	//CASE: List is empty
	if (L->n == 0) {
		printf("List Error: calling deleteFront() on an empty List\n");
		exit(1);
	}
	Node N = L->front;

	if (length(L) > 1) {

		L->front = L->front->next;
		L->front->prev = NULL;
		L->cursor = L->front;
	}
	else {
		L->front = L->back = NULL;
		L->cursor = L->front;
	}

	L->n--;
	freeNode(&N);
	N=NULL;
}

void deleteBack(List L) {

	//CASE: List is NULL
	if (L == NULL) {
		printf("List Error: calling deleteBack() on NULL List reference\n");
		exit(1);
	}
	//CASE: List is empty
	if (L->n == 0) {
		printf("List Error: calling deleteBack() on an empty List\n");
		exit(1);
	}

	Node N = L->back;

	if (length(L) > 1) {

		L->back = L->back->prev;
		L->back->next = NULL;
		L->cursor = L->back;
	}
	else {
		L->front = L->back = NULL;
		L->cursor = NULL;
	}

	L->n--;
	freeNode(&N);
	N=NULL;
}

void delete(List L)
{
	//CASE: List is NULL
	if (L == NULL) {
		printf("List Error: calling delete() on NULL List reference\n");
		exit(1);
	}
	//CASE: List is empty
	if (L->n == 0) {
		printf("List Error: calling delete() on an empty List\n");
		exit(1);
	}
	Node temp = L->cursor;
	if (L->n == 1) {
		L->front = L->back = NULL;
	}
	else {
		if (L->cursor == L->front) {
			L->front = L->front->next;
			L->front->prev = NULL;

		}
		else if (L->cursor == L->back) {
			L->back = L->back->prev;
			L->back->next = NULL;
		}
		else {
			L->cursor = L->cursor->prev;
			L->cursor->next = temp->next; // this line is broken
			L->cursor = temp->next;
			L->cursor->prev = temp->prev;
		}

	}

	L->cursor = NULL;
	freeNode(&temp);
	temp = NULL;


	L->index = -1;
	L->n--;
}

// Other operations -----------------------------------------------------------
void printList(FILE* out, List L) {

	if (L == NULL) {
		printf("List Error: calling printList() on NULL List reference\n");
		exit(1);
	}
	Node temp = L->cursor;
	int idx = index(L);
	for (moveFront(L); index(L) >= 0; moveNext(L)) {
		fprintf(out, "%d ", get(L));
	}
	L->cursor = temp;
	L->index = idx;
	//fprintf(out, "%s", "\n");
}

List copyList(List L) {

	//CASE: List is NULL
	if (L == NULL) {
		printf("List Error: calling deleteFront() on NULL List reference\n");
		exit(1);
	}
	//CASE: List is empty
	if (L->n == 0) {
		printf("List Error: calling deleteFront() on an empty List\n");
		exit(1);
	}

	List list = newList();

	for (moveFront(L); index(L) >= 0; moveNext(L))
		append(list, get(L));

	return(list);
}